### DISCUSSION TITLE
How to specify in_axes for custom pytree node?

### DISCUSSION THREAD

**Discussion by IrishWhiskey** [jax-ml/jax#16127] [Q&A] [answered]
Let's say I have a function, which takes a custom object as input, that I want to transform with pmap. How should the `in_axes` argument look like?

If I run
```
import jax
import jax.numpy as jnp
from functools import partial

@jax.tree_util.register_pytree_node_class
class C:
    def __init__(self, data, static_data):
        self.data = data
        self.static_data = static_data
        
    def tree_flatten(self):
        children = (self.data,)  # arrays / dynamic values
        aux_data = dict(static_data=self.static_data)  # static values
        return (children, aux_data)

    @classmethod
    def tree_unflatten(cls, aux_data, children):
        return cls(*children, **aux_data)

@partial(jax.pmap, in_axes=(C(0, None)))
def foo(x):
    return jnp.exp(x.data)

c = C(jnp.ones((100, 100)), 1)
foo(c)
```

I get `ValueError: pmap in_axes specification must be a tree prefix of the corresponding value, got specification (<__main__.C object at 0x7f26fedbd0c0>, 0) for value tree PyTreeDef(((CustomNode(C[{'static_data': 1}], [*]),), {})).`

Am I doing anything wrong? How can I fix it?

**Comment by patrick-kidger** [comment, marked as answer]
You want `in_axes=(C(0, 1),)`.

- `in_axes` should be a tuple of length equal to your number of input arguments.
- the static data for the custom pytree node must be the same.

**Reply by davisyoshida** [reply]
It's a bit unfortunate that this seems to rule out custom pytrees typechecking their inputs

**Reply by patrick-kidger** [reply]
Not really -- the unflatten rule just shouldn't use `__init__`. See [the unflatten rule](https://github.com/patrick-kidger/equinox/blob/566d02c15bb017dd69f5e5fb43f54787a5f52076/equinox/_module.py#L261) for `equinox.Module` as an example.

**Reply by davisyoshida** [reply]
The thing I was talking about wasn't the unflatten rule but the fact that you have to call `C(0, 1)`, so `__init__` needs to be happy being called with integers.

**Reply by patrick-kidger** [reply]
Ah, right. You can actually produce that object via unflattening, without needing to go via `__init__`, but I'll grant you it's not as tidy: `eqx.tree_at(lambda x: x.data, c, 0)`.

FWIW I've never actually needed to write something like `C(0, 1)` in practice. I nearly always start with with some object, and then just make edits to it:
```python
tree = ...   # some pytree
in_axes = jtu.tree_map(lambda _: 0, tree)  # map most leaves over their first axis
in_axes = eqx.tree_at(lambda t: t.layers[-1].weight, in_axes, 1)  # map `tree.layers[-1].weight` over the second axis
jax.vmap(..., in_axes=in_axes)(tree)
```
(This is why I focussed on the unflattening case, as it's what I primarily had in my head for constructing `in_axes`.)

**Reply by davisyoshida** [reply]
Makes sense. I think the fact that workarounds like this are necessary makes it not an ideal API though.

**Reply by jakevdp** [reply]
For what it's worth, the best mental model to have for pytrees is "a structure that maps to a tuple". This does preclude some more complicated pytree structures, but there is value in the simplicity of the model. In particular, your pytrees should generally avoid doing data validation on `__init__` for the reasons mentioned above (there's a bit of discussion of this in [Extending Pytrees](https://jax.readthedocs.io/en/latest/pytrees.html#extending-pytrees)). Various downstream libraries, (including Equinox, flax, haiku, etc.) have developed frameworks that build on or extend the capabilities of JAX's very simple pytree model, but the JAX team generally feels that, in the core library at least, a simpler framework better serves the needs of general downstream users.

**Reply by davisyoshida** [reply]
Thanks for the response, I'll keep that in mind. It's pretty hard to resist getting fancy with them given how useful they are though!

**Reply by darsnack** [reply]
Apologies for hijacking this thread after many months, but I've been searching for quite some time on this issue, and this answer is the closest discussion I've found.

So, following up on above, the requirement that static leaves are the same means that it isn't possible to specify `out_axes` without already having an instance of the object? This precludes vmapping class creation methods over seeds as an example.

**Reply by jakevdp** [reply]
I think that's correct: in order to specify non-trivial `out_axes`, you need to pass an instance of the class. If you're unable to construct an instance of the class for some reason, you will not be able to specify non-trivial `out_axes`. That said, I don't totally understand the context of what you describe: I'd suggest opening a new issue with a minimal repro of the kind of thing you have in mind.

**Reply by patrick-kidger** [reply]
Getting an instance of an object is doable by first calling `jax.eval_shape` :)

(And then modifying it with e.g. `jax.tree_map` and `equinox.tree_at` to obtain your desired `out_axes`.)

**Reply by jakevdp** [reply]
Whether `eval_shape` is a solution depends on the details of the setting the user describes. Let's move this discussion to a new issue if @darsnack would like to dig-in more.
